gencred/cmd/main.go

```diff
@@ -0,0 +1,219 @@
+	// defaultConfigFileName is the default kubeconfig filename.
+	defaultConfigFileName = "/dev/stdout"
+)
+
+// options are the available command-line flags.
+type options struct {
+	context        string
+	name           string
+	output         string
+	certificate    bool
+	serviceaccount bool
+	overwrite      bool
+}
+
+// exitError is a custom error type which stores a message and status code.
+type exitError struct {
+	code    int
+	message string
+}
+
+func (err exitError) Error() string {
+	return err.message
+}
+
+// printErrAndExit prints an error message to stderr and exits with a status code.
+func printErrAndExit(err error) {
+	_, _ = fmt.Fprintln(os.Stderr, err.Error())
+
+	exitErr, ok := err.(*exitError)
+	if ok {
+		os.Exit(exitErr.code)
+	} else {
+		os.Exit(1)
+	}
+}
+
+// parseFlags parses the command-line flags.
+func (o *options) parseFlags() {
+	// --context
+	flag.StringVar(&o.context, "context", "", "The name of the kubeconfig context to use.")
+
+	// -n, --name
+	flag.StringVarP(&o.name, "name", "n", defaultContextName, "Context name for the kubeconfig entry.")
+
+	// -o, --output
+	flag.StringVarP(&o.output, "output", "o", defaultConfigFileName, "Output path for generated kubeconfig file.")
+
+	// -c, --certificate
+	flag.BoolVarP(&o.certificate, "certificate", "c", false, "Authorize with a client certificate and key.")
+
+	// -s, --serviceaccount
+	flag.BoolVarP(&o.serviceaccount, "serviceaccount", "s", false, "Authorize with a service account.")
+
+	// --overwrite
+	flag.BoolVar(&o.overwrite, "overwrite", false, "Overwrite (rather than merge) output file if exists.")
+
+	flag.Parse()
+}
+
+// validateFlags validates the command-line flags.
+func (o *options) validateFlags() error {
+	var err error
+
+	if len(o.name) == 0 {
+		return &exitError{message: "-n, --name option is required.", code: 1}
+	}
+
+	o.output, err = filepath.Abs(o.output)
+	if err != nil {
+		return &exitError{message: fmt.Sprintf("-o, --output option invalid: %v.", o.output), code: 1}
+	}
+
+	if dirExists(o.output) {
+		return &exitError{message: fmt.Sprintf("-o, --output already exists and is a directory: %v.", o.output), code: 1}
+	}
+
+	if o.serviceaccount && o.certificate {
+		return &exitError{message: "-c, --certificate and -s, --serviceaccount are mutually exclusive options.", code: 1}
+	}
+
+	return nil
+}
+
+// dirExists checks if a path exists and is a directory.
+func dirExists(path string) bool {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return info.IsDir()
+}
+
+// fileExists checks if a path exists and is a regular file.
+func fileExists(path string) bool {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return info.Mode().IsRegular()
+}
+
+// mergeConfigs merges an existing kubeconfig file with a new entry with precedence given to the existing config.
+func mergeConfigs(o options, kubeconfig []byte) ([]byte, error) {
+	tmpFile, err := ioutil.TempFile("", "")
```

Comment:
Seems like it would be moderately useful to give it some comprehensible prefix

Reply:
Since this temp file is *throw-away* and needed only for the kubeconfig merging (and deleted afterward) a random string should be fine right?